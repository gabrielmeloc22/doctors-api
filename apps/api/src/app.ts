import bodyParser from "body-parser"
import cors from "cors"
import express, { type Express } from "express"
import morgan from "morgan"

export const createApp = (): Express => {
    const app = express()

    app.disable("x-powered-by")
        .use(morgan("dev"))
        .use(bodyParser.urlencoded({ extended: true }))
        .use(bodyParser.json())
        .use(cors())

    return app
}
