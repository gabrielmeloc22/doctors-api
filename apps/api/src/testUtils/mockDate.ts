import { vi } from "vitest";

export const mockDate = (date: Date): void => {
    vi.useFakeTimers();
    vi.setSystemTime(date);
};
