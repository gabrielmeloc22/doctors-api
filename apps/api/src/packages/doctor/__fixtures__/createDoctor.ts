import { db } from "../../../database/db";
import { DeepPartial } from "../../../utils/deepPartial";
import { doctorTable, IDoctor } from "../doctorTable";

export const createDoctor = async (
    args?: DeepPartial<IDoctor>
): Promise<IDoctor> => {
    const doctorCount = await db.$count(doctorTable);

    const [doctor] = await db
        .insert(doctorTable)
        .values({
            email: `doc${doctorCount}@mail.com`,
            firstName: "john",
            lastName: `doe ${doctorCount}`,
            username: `john#${doctorCount}`,
            ...args,
        })
        .returning();

    return doctor as IDoctor;
};
