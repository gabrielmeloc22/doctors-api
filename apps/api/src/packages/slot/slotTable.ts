import * as t from "drizzle-orm/pg-core";
import { idField, timestampField } from "../database/schemaDefaults";
import { doctorTable } from "../doctor/doctorTable";

export const slotTable = t.pgTable("slot", {
    doctorId: t.uuid().references(() => doctorTable.id),
    startTime: t.date().notNull(),
    endTime: t.date().notNull(),
    available: t.boolean(),
    ...timestampField,
    ...idField,
});
