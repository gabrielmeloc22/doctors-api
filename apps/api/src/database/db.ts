import { drizzle } from "drizzle-orm/node-postgres";
import * as booking from "../packages/booking/bookingTable";
import * as doctor from "../packages/doctor/doctorTable";
import * as slot from "../packages/slot/slotTable";

export const schema = {
    ...slot,
    ...doctor,
    ...booking,
};

export const db = drizzle({
    connection: process.env.DATABASE_URL,
    casing: "camelCase",
    schema,
});
