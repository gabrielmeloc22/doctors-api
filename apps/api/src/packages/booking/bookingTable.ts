import * as t from "drizzle-orm/pg-core";
import { idField, timestampField } from "../../database/schemaDefaults";
import { slotTable } from "../slot/slotTable";

export const bookingTable = t.pgTable(
    "booking",
    {
        slotId: t
            .uuid()
            .references(() => slotTable.id)
            .notNull(),
        patientId: t.varchar({ length: 256 }).notNull(),
        reason: t.varchar({ length: 256 }).notNull(),
        time: t.timestamp().notNull(),
        ...timestampField,
        ...idField,
    },
    (table) => [
        {
            slot_id_time_unique: t.unique().on(table.id, table.time),
        },
    ]
);
