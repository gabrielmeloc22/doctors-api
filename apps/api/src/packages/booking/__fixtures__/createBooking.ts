import { db } from "../../../database/db";
import { DeepPartial } from "../../../utils/deepPartial";
import { createSlot } from "../../slot/__fixtures__/createSlot";
import { bookingTable, IBooking } from "../bookingTable";

export const createBooking = async (
    args?: DeepPartial<IBooking>
): Promise<IBooking> => {
    const slotId = args?.slotId ? args.slotId : (await createSlot()).id;

    const [booking] = await db
        .insert(bookingTable)
        .values({
            slotId: slotId,
            time: new Date(),
            patientId: "1",
            reason: "checkup",
            ...args,
        })
        .returning();

    return booking as IBooking;
};
