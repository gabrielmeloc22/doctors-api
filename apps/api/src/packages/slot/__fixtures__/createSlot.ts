import { add } from "date-fns";
import { db } from "../../../database/db";
import { DeepPartial } from "../../../utils/deepPartial";
import { createDoctor } from "../../doctor/__fixtures__/createDoctor";
import { ISlot, SLOT_REPEAT_TYPE_ENUM, slotTable } from "../slotTable";

export const createSlot = async (args?: DeepPartial<ISlot>): Promise<ISlot> => {
    const doctor = await createDoctor();

    const [slot] = await db
        .insert(slotTable)
        .values({
            doctorId: doctor.id,
            duration: 30,
            startTime: new Date(),
            endTime: add(new Date(), { minutes: 30 }),
            repeatType: SLOT_REPEAT_TYPE_ENUM.SINGLE,
            ...args,
        })
        .returning();

    return slot as ISlot;
};
