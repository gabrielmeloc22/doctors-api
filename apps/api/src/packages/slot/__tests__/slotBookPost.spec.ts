import request from "supertest";
import { beforeAll, expect, it } from "vitest";
import { createApp } from "../../../app";
import { db } from "../../../database/db";
import { mockDate } from "../../../testUtils/mockDate";
import {
    defaultFrozenKeys,
    sanitizeTestObject,
} from "../../../testUtils/sanitizeTestObject";
import { bookingTable } from "../../booking/bookingTable";
import { createDoctor } from "../../doctor/__fixtures__/createDoctor";
import { createSlot } from "../__fixtures__/createSlot";
import { SLOT_REPEAT_TYPE_ENUM } from "../slotTable";

beforeAll(() => {
    mockDate(new Date("2024-12-21T10:00:00.000Z"));
});

it("should validate params and return 400", async () => {
    const app = createApp();

    const payload = {
        reason: "checkup",
        patient_id: "some-id",
        start_time: new Date("2024-12-21T10:30:00.000Z").toISOString(),
    };

    const response = await request(app.listen())
        .post(`/slots/1/book`)
        .send(payload);

    expect(response.status).toBe(400);

    const body = response.body as Record<string, unknown>;

    expect(body).toMatchSnapshot();
});

it("should validate body and return 400", async () => {
    const app = createApp();

    const payload = {
        start_time: "",
    };

    const response = await request(app.listen())
        .post("/slots/0a523035-fcca-4bbd-a7e6-3918e4689f56/book")
        .send(payload);

    expect(response.status).toBe(400);

    const body = response.body as Record<string, unknown>;

    expect(body).toMatchSnapshot();
});

it("should validate slot id and return 404", async () => {
    const app = createApp();

    const payload = {
        reason: "checkup",
        patient_id: "some-id",
        start_time: new Date("2024-12-21T10:30:00.000Z").toISOString(),
    };

    const response = await request(app.listen())
        .post("/slots/0a523035-fcca-4bbd-a7e6-3918e4689f56/book")
        .send(payload);

    const body = response.body as Record<string, unknown>;

    expect(response.status).toBe(404);

    expect(body).toMatchObject({
        success: false,
        error: "Slot not found",
    });
});

it("should post, book a in a single slot and return 201", async () => {
    const app = createApp();

    const doctor = await createDoctor();
    const slot = await createSlot({
        duration: 30,
        doctorId: doctor.id,
        repeatType: SLOT_REPEAT_TYPE_ENUM.SINGLE,
        startTime: new Date("2024-12-21T10:00:00.000Z"),
        endTime: new Date("2024-12-21T12:00:00.000Z"),
    });

    const payload = {
        reason: "checkup",
        patient_id: "some-id",
        start_time: new Date("2024-12-21T10:30:00.000Z").toISOString(),
    };

    const response = await request(app.listen())
        .post(`/slots/${slot.id}/book`)
        .send(payload);

    const body = response.body as Record<string, unknown>;

    const bookings = await db.select().from(bookingTable);

    const [booking] = bookings;

    expect(response.status).toBe(201);

    expect(bookings).toHaveLength(1);

    expect(body).toMatchObject({
        success: true,
        booking: {
            id: booking?.id,
            patient_id: booking?.patientId,
            reason: booking?.reason,
            slot_id: booking?.slotId,
            time: booking?.time.toISOString(),
        },
    });

    expect(
        sanitizeTestObject({
            obj: body,
            frozenKeys: [...defaultFrozenKeys, "slot_id"],
        })
    ).toMatchSnapshot();
});

it("should post with unavailable date in a single slot and return 400", async () => {
    const app = createApp();

    const doctor = await createDoctor();
    const slot = await createSlot({
        duration: 30,
        doctorId: doctor.id,
        repeatType: SLOT_REPEAT_TYPE_ENUM.SINGLE,
        startTime: new Date("2024-12-21T10:00:00.000Z"),
        endTime: new Date("2024-12-21T12:00:00.000Z"),
    });

    const payload = {
        reason: "checkup",
        patient_id: "some-id",
        start_time: new Date("2024-12-20T12:00:00.000Z").toISOString(),
    };

    const response = await request(app.listen())
        .post(`/slots/${slot.id}/book`)
        .send(payload);

    const body = response.body as Record<string, unknown>;

    const bookings = await db.select().from(bookingTable);

    expect(response.status).toBe(400);

    expect(bookings).toHaveLength(0);

    expect(body).toMatchObject({
        success: false,
        error: "Booking time not available",
    });
});

it("should post, book a in a daily slot and return 201", async () => {
    const app = createApp();

    const doctor = await createDoctor();
    const slot = await createSlot({
        duration: 30,
        doctorId: doctor.id,
        repeatType: SLOT_REPEAT_TYPE_ENUM.DAILY,
        startTime: new Date("2024-12-21T10:00:00.000Z"),
        endTime: new Date("2024-12-21T12:00:00.000Z"), // 4 time slots
    });

    const payload = {
        reason: "checkup",
        patient_id: "some-id",
        start_time: new Date("2024-12-22T10:00:00.000Z").toISOString(),
    };

    const response = await request(app.listen())
        .post(`/slots/${slot.id}/book`)
        .send(payload);

    const body = response.body as Record<string, unknown>;

    const bookings = await db.select().from(bookingTable);

    const [booking] = bookings;

    expect(response.status).toBe(201);

    expect(bookings).toHaveLength(1);

    expect(body).toMatchObject({
        success: true,
        booking: {
            id: booking?.id,
            patient_id: booking?.patientId,
            reason: booking?.reason,
            slot_id: booking?.slotId,
            time: booking?.time.toISOString(),
        },
    });

    expect(
        sanitizeTestObject({
            obj: body,
            frozenKeys: [...defaultFrozenKeys, "slot_id"],
        })
    ).toMatchSnapshot();
});

it("should post with unavailable date in a daily slot and return 400", async () => {
    const app = createApp();

    const doctor = await createDoctor();
    const slot = await createSlot({
        duration: 30,
        doctorId: doctor.id,
        repeatType: SLOT_REPEAT_TYPE_ENUM.DAILY,
        startTime: new Date("2024-12-21T10:00:00.000Z"),
        endTime: new Date("2024-12-21T12:00:00.000Z"),
    });

    const payload = {
        reason: "checkup",
        patient_id: "some-id",
        start_time: new Date("2024-12-21T13:00:00.000Z").toISOString(),
    };

    const response = await request(app.listen())
        .post(`/slots/${slot.id}/book`)
        .send(payload);

    const body = response.body as Record<string, unknown>;

    const bookings = await db.select().from(bookingTable);

    expect(response.status).toBe(400);

    expect(bookings).toHaveLength(0);

    expect(body).toMatchObject({
        success: false,
        error: "Booking time not available",
    });
});

it("should post, book a in a weekly slot and return 201", async () => {
    const app = createApp();

    const doctor = await createDoctor();
    const slot = await createSlot({
        duration: 30,
        doctorId: doctor.id,
        repeatType: SLOT_REPEAT_TYPE_ENUM.WEEKLY,
        startTime: new Date("2024-12-21T10:00:00.000Z"),
        endTime: new Date("2024-12-21T12:00:00.000Z"), // 4 time slots
        repeatWeekdays: [1, 3, 5], // monday, wednesday, friday
    });

    const payload = {
        reason: "checkup",
        patient_id: "some-id",
        start_time: new Date("2024-12-23T10:00:00.000Z").toISOString(),
    };

    const response = await request(app.listen())
        .post(`/slots/${slot.id}/book`)
        .send(payload);

    const body = response.body as Record<string, unknown>;

    const bookings = await db.select().from(bookingTable);

    const [booking] = bookings;

    expect(response.status).toBe(201);

    expect(bookings).toHaveLength(1);

    expect(body).toMatchObject({
        success: true,
        booking: {
            id: booking?.id,
            patient_id: booking?.patientId,
            reason: booking?.reason,
            slot_id: booking?.slotId,
            time: booking?.time.toISOString(),
        },
    });

    expect(
        sanitizeTestObject({
            obj: body,
            frozenKeys: [...defaultFrozenKeys, "slot_id"],
        })
    ).toMatchSnapshot();
});

it("should post with unavailable date in a weekly slot and return 400", async () => {
    const app = createApp();

    const doctor = await createDoctor();
    const slot = await createSlot({
        duration: 30,
        doctorId: doctor.id,
        repeatType: SLOT_REPEAT_TYPE_ENUM.WEEKLY,
        startTime: new Date("2024-12-21T10:00:00.000Z"),
        endTime: new Date("2024-12-21T12:00:00.000Z"), // 4 time slots
        repeatWeekdays: [1, 3, 5], // monday, wednesday, friday
    });

    const payload = {
        reason: "checkup",
        patient_id: "some-id",
        start_time: new Date("2024-12-22T10:00:00.000Z").toISOString(),
    };

    const response = await request(app.listen())
        .post(`/slots/${slot.id}/book`)
        .send(payload);

    const body = response.body as Record<string, unknown>;

    const bookings = await db.select().from(bookingTable);

    expect(response.status).toBe(400);

    expect(bookings).toHaveLength(0);

    expect(body).toMatchObject({
        success: false,
        error: "Booking time not available",
    });
});

it("should post with unavailable time in a weekly slot and return 400", async () => {
    const app = createApp();

    const doctor = await createDoctor();
    const slot = await createSlot({
        duration: 30,
        doctorId: doctor.id,
        repeatType: SLOT_REPEAT_TYPE_ENUM.WEEKLY,
        startTime: new Date("2024-12-21T10:00:00.000Z"),
        endTime: new Date("2024-12-21T12:00:00.000Z"), // 4 time slots
        repeatWeekdays: [1, 3, 5], // monday, wednesday, friday
    });

    const payload = {
        reason: "checkup",
        patient_id: "some-id",
        start_time: new Date("2024-12-23T13:00:00.000Z").toISOString(),
    };

    const response = await request(app.listen())
        .post(`/slots/${slot.id}/book`)
        .send(payload);

    const body = response.body as Record<string, unknown>;

    const bookings = await db.select().from(bookingTable);

    expect(response.status).toBe(400);

    expect(bookings).toHaveLength(0);

    expect(body).toMatchObject({
        success: false,
        error: "Booking time not available",
    });
});
