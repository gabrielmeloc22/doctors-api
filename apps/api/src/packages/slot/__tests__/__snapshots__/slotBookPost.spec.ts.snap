// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should post, book a in a daily slot and return 201 1`] = `
{
  "booking": {
    "id": "FROZEN-ID",
    "patient_id": "some-id",
    "reason": "checkup",
    "slot_id": "FROZEN-SLOT_ID",
    "time": "2024-12-22T10:00:00.000Z",
  },
  "success": true,
}
`;

exports[`should post, book a in a single slot and return 201 1`] = `
{
  "booking": {
    "id": "FROZEN-ID",
    "patient_id": "some-id",
    "reason": "checkup",
    "slot_id": "FROZEN-SLOT_ID",
    "time": "2024-12-21T10:30:00.000Z",
  },
  "success": true,
}
`;

exports[`should post, book a in a weekly slot and return 201 1`] = `
{
  "booking": {
    "id": "FROZEN-ID",
    "patient_id": "some-id",
    "reason": "checkup",
    "slot_id": "FROZEN-SLOT_ID",
    "time": "2024-12-23T10:00:00.000Z",
  },
  "success": true,
}
`;

exports[`should validate body and return 400 1`] = `
{
  "error": [
    {
      "message": "Required",
      "path": [
        "reason",
      ],
    },
    {
      "message": "Required",
      "path": [
        "patient_id",
      ],
    },
    {
      "message": "Invalid datetime",
      "path": [
        "start_time",
      ],
    },
  ],
  "success": false,
}
`;

exports[`should validate params and return 400 1`] = `
{
  "error": [
    {
      "message": "Invalid uuid",
      "path": [
        "id",
      ],
    },
  ],
  "success": false,
}
`;
